test/test_tagging.py

```diff
@@ -0,0 +1,235 @@
+import json
+from typing import (
+    Optional,
+)
+import unittest
+from unittest.mock import (
+    MagicMock,
+    patch,
+)
+
+from azul.deployment import (
+    populate_tags,
+)
+from azul.logging import (
+    configure_test_logging,
+)
+from azul_test_case import (
+    AzulUnitTestCase,
+)
+
+
+def setupModule():
+    configure_test_logging()
+
+
+class TestTerraformResourceTags(AzulUnitTestCase):
+
+    def assertDictEqualPermissive(self, expected, actual):
+        """
+        Assert wrapper for permissive_compare.
+        """
+        path = self.permissive_compare(expected, actual)
+        return self.assertIsNone(path, f'Discrepancy in path: {path}')
+
+    def permissive_compare(self, expected, actual, *path: str) -> Optional[str]:
+        """
+        Recursive dictionary comparison. Skips comparing values when the
+        reference value is set to None.
+
+        Returns None if dictionaries match, or path of discrepancy as a
+        string otherwise.
+
+        >>> t = TestTerraformResourceTags()
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 789}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 789}]}
+        ... )
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: None}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+
+        >>> t.permissive_compare(  #doctest: +ELLIPSIS
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'def'}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        ...
+        'qaz[0].456'
+
+        >>> t.permissive_compare(  #doctest: +ELLIPSIS
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'def'}]},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: None}]}
+        ... )
+        'qaz[0].456'
+
+        >>> t.permissive_compare(  #doctest: +ELLIPSIS
+        ...     {'foo': 'bar', 'qaz': None, 456: 'def'},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        'object length mismatch'
+
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': None},
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        """
+        # in case actual is a superset of expected
+        if len(actual) != len(expected):
+            return 'object length mismatch'
```

Comment:
This seems weird. This should return the path, no? Maybe the function should return the path *and* a string explaining the difference. But just the path is fine; at the root that would be the empty tuple.